### DISCUSSION TITLE
Production site has issues with search and theme switcher.

### DISCUSSION THREAD

**Discussion by johnnymatthews** [thuliteio/doks#718] [Q&A]
When running the site locally with `npm run start`, everything works as expected. However, when building the site with `npm run build` and then either throwing that on a server (Netlify, Fleek, etc) or running a local server with `serve`, neither the search or the dark/light mode switcher works. Here's the repo and branch that's having issues: https://github.com/filecoin-project/filecoin-docs/tree/doks-theme

Reading other comments and discussions, it seems like this has _something_ to do with the `baseurl` variable in `_default/config.toml` and `production/config.toml`, but I've tried a bunch of different combinations and nothing seems to work.

Oddly enough, [this other repo](https://github.com/filecoin-project/lotus-docs) has `baseurl` in `_default/config.toml` set to `"/"` and everything works fine! So I'm not quite sure what's going on. There _could_ be some kind of weird caching thing happening? But I'm pretty sure [using `serve`](https://www.npmjs.com/package/serve) would get around this issue.

Sorry for the wall of text. Any ideas?

**Comment by h-enk** [comment]
My advice:
- ONLY set your production URL in [`./config/_default/config.toml`](https://github.com/h-enk/doks/blob/master/config/_default/config.toml#L1) — so, do NOT set any URL in [`./config/next/config.toml`](https://github.com/h-enk/doks/blob/master/config/next/config.toml) or [`./config/production/config.toml`](https://github.com/h-enk/doks/blob/master/config/production/config.toml). Then, Hugo, Netlify and others can perfectly make out the correct URL to use.
- Don't use [Hugo server](https://gohugo.io/commands/hugo_server/) (it's limited), but instead use e.g. [http-server](https://github.com/http-party/http-server) or [Caddy](https://github.com/caddyserver/caddy).

I got no issues when I did this:

1. `git clone -b doks-theme --single-branch git@github.com:filecoin-project/filecoin-docs.git filecoin-docs`
2. `cd filecoin-docs && npm i`
3. `npx exec-bin node_modules/.bin/hugo/hugo --gc --minify --baseURL=http://192.168.1.10:8080`
4. `http-server`

**Reply by johnnymatthews** [reply]
Thanks for the reply Henk! Sadly, I'm still getting the same issue, even when I followed your instructions (literally just copy and pasted your commands). Don't know how useful this paste will be, but here's exactly what happened in my terminal:

```
$ › git clone -b doks-theme --single-branch git@github.com:filecoin-project/filecoin-docs.git filecoin-docs
cd filecoin-docs && npm i
npx exec-bin node_modules/.bin/hugo/hugo --gc --minify --baseURL=http://192.168.1.10:8080
http-server
Cloning into 'filecoin-docs'...
Enter passphrase for key '/Users/johnny/.ssh/id_ed25519':
remote: Enumerating objects: 16866, done.
remote: Counting objects: 100% (5605/5605), done.
remote: Compressing objects: 100% (1511/1511), done.
remote: Total 16866 (delta 4487), reused 4864 (delta 3963), pack-reused 11261
Receiving objects: 100% (16866/16866), 44.00 MiB | 26.76 MiB/s, done.
Resolving deltas: 100% (11677/11677), done.
npm WARN deprecated @stylelint/postcss-markdown@0.36.2: Use the original unforked package instead: postcss-markdown
npm WARN deprecated @braintree/sanitize-url@3.1.0: Potential XSS vulnerability patched in v6.0.0.

> @hyas/doks@0.3.4 postinstall
> hugo-installer --version otherDependencies.hugo --extended --destination node_modules/.bin/hugo


Hugo Installer

> Checking for existing binary
> Downloading binary from "https://github.com/gohugoio/hugo/releases/download/v0.88.1/hugo_extended_0.88.1_macOS-64bit.tar.gz"
> Downloading checksum from "https://github.com/gohugoio/hugo/releases/download/v0.88.1/hugo_0.88.1_checksums.txt"
> Verifying binary checksum
> Extracting binary to disk
> Verifying binary health

Hugo is now available in "node_modules/.bin/hugo".

- Version       0.88.1
- Extended      Yes
- OS            darwin
- Architecture  x64

hugo v0.88.1-5BC54738+extended darwin/amd64 BuildDate=2021-09-04T09:39:19Z VendorInfo=gohugoio

Success!

added 752 packages, and audited 753 packages in 9s

99 packages are looking for funding
  run `npm fund` for details

5 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
Start building sites …
hugo v0.88.1-5BC54738+extended darwin/amd64 BuildDate=2021-09-04T09:39:19Z VendorInfo=gohugoio

                   | EN
-------------------+------
  Pages            |  76
  Paginator pages  |   0
  Non-page files   |  69
  Static files     | 100
  Processed images |   0
  Aliases          |   8
  Sitemaps         |   1
  Cleaned          |   0

Total in 2719 ms
Starting up http-server, serving ./public

http-server version: 14.1.0

http-server settings:
CORS: disabled
Cache: 3600 seconds
Connection Timeout: 120 seconds
Directory Listings: visible
AutoIndex: visible
Serve GZIP Files: false
Serve Brotli Files: false
Default File Extension: none

Available on:
  http://127.0.0.1:8080
  http://10.0.0.182:8080
  http://10.1.36.151:8080
Hit CTRL-C to stop the server
```

Unfortunately, search still doesn't work, and the dark/light toggle is non functional. Here's what the web console is showing me:

<img width="1440" alt="Screen Shot 2022-03-30 at 11 36 02 AM" src="https://user-images.githubusercontent.com/9611008/160874429-746a4848-c6aa-4416-bbdc-7d69a15f9da1.png">

There's a chance that something has gone wrong somewhere else in the development of this site. So I'm gonna try rolling back commits until I get a working version. Thanks for your help all the same! I'll post an answer here for other folks to use if I find anything.

**Reply by h-enk** [reply]
> npx exec-bin node_modules/.bin/hugo/hugo --gc --minify --baseURL=http://192.168.1.10:8080

Is baseURL the local IP address of your machine?

**Reply by johnnymatthews** [reply]
According to `ipconfig getifaddr en0` my IP is `10.0.0.182`. So I ran `npx exec-bin node_modules/.bin/hugo/hugo --gc --minify --baseURL=http://10.0.0.182:8080` but still got the same errors in the dev browser console. Also tested in a few different browsers but got the same issues.

Just to see if it's a local issue, I build the site using `npx exec-bin node_modules/.bin/hugo/hugo --gc --minify` and created a new site on Netlify by dragging over the `public` folder. The [preview is here[(https://rainbow-seahorse-898076.netlify.app/) but still has the same issue. In this version I removed the `baseurl` variable from `config/next/config.json` and `config/production/config.json`, and set the `baseurl` to `"/"` in `config/_default/config.json`. The preview still has the same issues, so I think there's something else funky going on.

**Reply by h-enk** [reply]
> According to ipconfig getifaddr en0 my IP is 10.0.0.182.

Run `ipconfig`, and pick the `IPv4 Address` of (probably) the first adapter — it should (most likely) start w/ `192` (not `10`)

> Just to see if it's a local issue, I build the site using npx exec-bin node_modules/.bin/hugo/hugo --gc --minify and created a new site on Netlify by dragging over the public folder. 

That's handy! You have Content Security Policy issues (that prevent loading JavaScript files and are breaking the site). You'll first need to fix them in `./layouts/index.headers`. See also [Content Security Policy](https://gethyas.com/docs/reference-guides/security/#content-security-policy) (I will add this to getdoks.org)
 
What I see:

- Add https://www.googletagmanager.com/gtag/js + `sha256-5CW8UlkHXETe0M1q6Yr9K0txvX7lTHK4mQaW8/7KgAY=` to `script-src`
- Add https://www.google-analytics.com to `connect-src`
- Add https://www.googletagmanager.com to `img-src`

Next to that, you have issues with your [index.js](https://github.com/h-enk/doks/blob/master/assets/js/index.js](https://github.com/filecoin-project/filecoin-docs/blob/doks-theme/assets/js/index.js)) file. Could you try uncommenting [mainSections](https://github.com/filecoin-project/filecoin-docs/blob/doks-theme/config/_default/params.toml#L10)?

![Snag_88d4195](https://user-images.githubusercontent.com/3902872/161007140-e30727f4-9f15-4ca0-9b04-999d0cc06c7d.png)

**Reply by FactoryOptic** [reply]
> My advice:
> 
> * ONLY set your production URL in [`./config/_default/config.toml`](https://github.com/h-enk/doks/blob/master/config/_default/config.toml#L1) — so, do NOT set any URL in [`./config/next/config.toml`](https://github.com/h-enk/doks/blob/master/config/next/config.toml) or [`./config/production/config.toml`](https://github.com/h-enk/doks/blob/master/config/production/config.toml). Then, Hugo, Netlify and others can perfectly make out the correct URL to use.
`

This conflicts with the [documentation](https://getdoks.org/docs/tutorial/deploy-your-site/)

`Make sure to set baseURL in ./config/production/config.toml to the URL of your Doks website in production.`

**Reply by h-enk** [reply]
Thanks, the documentation needs updating for my advice: https://github.com/h-enk/getdoks.org/issues/294

**Reply by qorncat** [reply]
the action of making or manufacturing from components or raw materials, or the process of being so manufactured.

**Reply by sivanov-nuodb** [reply]
I have similar issues with FlexSearch when the `baseUrl` contains a path in the URI (e.g. `http://localhost:8080/nuodb-cp-docs/`). The `search-index.json` is still requested from the incorrect base path. Is there something that I'm missing here?

Steps to reproduce:

1 . Build the site

```sh
rm -rf public tmp/public
npm run build -- --baseURL http://localhost:8080/nuodb-cp-docs/
```

2. Serve the generated site

```sh
cp public tmp/nuodb-cp-docs
cd tmp && http-server
```
Navigate to `http://localhost:8080/nuodb-cp-docs/` in the browser and try to search.
The below log shows the 404 errors preventing search from working.

```
[Mon Aug 19 2024 18:07:31 GMT+0300 (Eastern European Summer Time)]  "GET /search-index.json" Error (404): "Not found"
[Mon Aug 19 2024 18:07:31 GMT+0300 (Eastern European Summer Time)]  "GET /favicon-192x192.png" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36"
[Mon Aug 19 2024 18:07:31 GMT+0300 (Eastern European Summer Time)]  "GET /favicon-192x192.png" Error (404): "Not found"
```

**Reply by h-enk** [reply]
See https://github.com/thuliteio/doks/issues/1275
